src/main/webapp/app/exercises/text/assess/text-submission-assessment.component.ts

```diff
@@ -227,6 +238,88 @@ export class TextSubmissionAssessmentComponent extends TextAssessmentBaseCompone
+                newTextBlockRefs.push(existingBlockRef);
+            } else {
+                if (exStart < start) {
+                    // Existing text block starts before text block to add
+                    if (exEnd > end) {
+                        // Existing text block ends after text block to add => covers the whole text block to add
+                        // Split the existing text block into two text blocks
+                        const newBlockRef = new TextBlockRef(new TextBlock(), undefined);
+                        newBlockRef.block!.startIndex = end;
+                        newBlockRef.block!.endIndex = exEnd;
+                        newBlockRef.block!.submission = this.submission;
+                        newBlockRef.block!.computeId();
+
+                        existingBlockRef.block!.endIndex = start;
+                        newTextBlockRefs.push(existingBlockRef);
+                        newTextBlockRefs.push(newBlockRef);
+                    } else {
+                        existingBlockRef.block!.endIndex = start;
+                        newTextBlockRefs.push(existingBlockRef);
+                    }
+                } else if (exEnd > end) {
+                    // Existing text block ends after text block to add
+                    existingBlockRef.block!.startIndex = end;
+                    newTextBlockRefs.push(existingBlockRef);
+                }
+            }
+        }
+
+        // Add the text block to add
+        newTextBlockRefs.push(refToAdd);
+
+        // Sort the new text block refs by their start index
+        newTextBlockRefs.sort((ref1, ref2) => ref1.block!.startIndex! - ref2.block!.startIndex!);
+
+        // Update the text on all text block refs
+        for (const blockRef of newTextBlockRefs) {
+            blockRef.block!.text = this.submission!.text!.substring(blockRef.block!.startIndex!, blockRef.block!.endIndex!);
+        }
+
+        this.textBlockRefs = newTextBlockRefs;
+        this.submission!.blocks = this.textBlockRefs.map((blockRef) => blockRef.block!);
+        this.result!.feedbacks = this.textBlockRefs.map((blockRef) => blockRef.feedback).filter((feedback) => feedback != undefined) as Feedback[];
+    }
+
+    /**
+     * Start loading feedback suggestions from Athena
+     * (only if this is a fresh submission, i.e. no assessments exist yet)
+     */
+    loadFeedbackSuggestions(): void {
+        if (this.assessments.length > 0) {
+            return;
+        }
+        this.feedbackSuggestionsObservable = this.athenaService.getFeedbackSuggestions(this.exercise!.id!, this.submission!.id!).pipe(
+            map((feedbackSuggestions: TextBlockRef[]) => {
+                for (const suggestion of feedbackSuggestions) {
+                    this.addAutomaticTextBlockRef(suggestion);
+                }
+            }),
+        );
+        // actually consume data
+        this.feedbackSuggestionsObservable.subscribe();
```

Comment:
You save this observable but never destroy it. Afaik (and I'm absolutely no expert on this, please correct me if I'm wrong) we need to manually unsubscribe from these observables when the component gets destroyed, this could otherwise lead to memory leaks.

Reply:
I'm also unsure if this would have any implications, but to be sure, I now destroy the subscription onDestroy.